### PR TITLE
[WIP]Custom devices profile mode

### PR DESCRIPTION
**Pull request by pieter-scholtz**
This PR is adding profiling mode functionality to custom-devices , it extends flutter_tools to include **libflutter_engine.so** as an artifact, which enables inclusion of **libflutter_engine.so** to the asset bundle. As part of this PR custom devices build steps are added as Targets. 

This PR fixes the lack of being able to easily run in profiling mode on custom devices.

### CODE REVIEW

File: packages/flutter_tools/lib/src/artifacts.dart [hunk 1]
```diff
@@ -17,6 +17,8 @@ import 'globals.dart' as globals;
 enum Artifact {
   /// The tool which compiles a dart kernel file into native code.
   genSnapshot,
+  /// Hannes might know what call this.
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/artifacts.dart:20]: was this a TODO to check with Hannes?

File: packages/flutter_tools/lib/src/artifacts.dart [hunk 2]
```diff
@@ -941,7 +951,10 @@ class CachedLocalEngineArtifacts implements LocalEngineArtifacts {
   }
 
   String _genSnapshotPath() {
-    const List<String> clangDirs = <String>['.', 'clang_x64', 'clang_x86', 'clang_i386', 'clang_arm64'];
+    // Does this  new implementation break anything ?
+    // Can we work TargetPlatform and EnvironmentType into this conditionally somehow ?
+    // Will the follwoing break anything: return the gen_snapshot inside clang_x64 directory if targetPlatform == linux_arm64
+    const List<String> clangDirs = <String>['clang_x64', 'clang_x86', 'clang_i386', 'clang_arm64','.'];
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/artifacts.dart:957]: why did you move `.` to the end here?

File: packages/flutter_tools/lib/src/artifacts.dart [hunk 3]
```diff
@@ -952,6 +965,15 @@ class CachedLocalEngineArtifacts implements LocalEngineArtifacts {
     throw Exception('Unable to find $genSnapshotName');
   }
 
+  String _libFlutterEnginePath() {
+    final String libFlutterEngineName = _artifactToFileName(Artifact.libFlutterEngine)!;
+    final String libFlutterEnginePath = _fileSystem.path.join(engineOutPath, libFlutterEngineName);
+      if (_processManager.canRun(libFlutterEnginePath)) {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/artifacts.dart:971]: we won't be running this file, right?

File: packages/flutter_tools/lib/src/artifacts.dart [hunk 4]
```diff
@@ -162,6 +164,8 @@ String? _artifactToFileName(Artifact artifact, [ TargetPlatform? platform, Build
   switch (artifact) {
     case Artifact.genSnapshot:
       return 'gen_snapshot';
+    case Artifact.libFlutterEngine:
+      return 'libflutter_engine.so';
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/artifacts.dart:168]: Is this file part of the officially pre-built Flutter SDK, or only generated via a custom device engine build?

For example, on x64 linux I have `libflutter_linux_gtk.so`.

File: packages/flutter_tools/lib/src/artifacts.dart (hunk 1 above)
Comment (pieter-scholtz) [reply packages/flutter_tools/lib/src/artifacts.dart:20]: I am unsure if libFlutterEngine is the correct/best name for this , since it might be named differently on different platforms , at the time of writing I thought that Hannes might have a better suggestion for a name.

File: packages/flutter_tools/lib/src/artifacts.dart (hunk 4 above)
Comment (pieter-scholtz) [reply packages/flutter_tools/lib/src/artifacts.dart:168]: I hope I am grasping your question fully. 

**To give some background first:**

If build the flutter engine locally with `./flutter/tools/gn --target-os linux --linux-cpu x64` then the `out/linux_debug_x64` folder has `libflutter_engine.so` and `libflutter_linux_gtk.so` , I am not 100% sure what the difference between the two is but I do know that having`libflutter_engine.so` available is very convenient when working with a custom device. 

I strongly suspect that `libflutter_engine.so` is "always" generated when building the engine. 

My python is not great but from what I read [here ](https://flutter.googlesource.com/recipes/+/refs/heads/master/recipes/engine.py) in the function starting on ln754 `libflutter_engine.so` is plainly not copied to the hosted artifacts folder.

This pull request was written with locally built engines in mind. My reasoning is that working with a custom device would imply that you are building your own engine. Since if it was hosted by infra it would not be a custom device. 

**Answer:**

As far as my knowledge goes `libflutter_linux_gtk.so` would be downloaded from infra and is not part of the SDK. 

Assuming that officially pre-built flutter SDK is referring to engine build artifacts hosted infra, my answer is : No , it not part of the officially pre-built Flutter SDK , and is generated via a linux engine build.

File: packages/flutter_tools/lib/src/artifacts.dart (hunk 2 above)
Comment (pieter-scholtz) [reply packages/flutter_tools/lib/src/artifacts.dart:957]: Because I don't want to grab the `gen_snapshot` in the root engine build output folder , but rather the one in the 'clang_arm64' directory.

File: packages/flutter_tools/lib/src/artifacts.dart (hunk 3 above)
Comment (pieter-scholtz) [reply packages/flutter_tools/lib/src/artifacts.dart:971]: In practice on my machine _processManager.canRun() effectively just checks if the file exists, [here](https://github.com/google/process.dart/blob/master/lib/src/interface/local_process_manager.dart) ln124 and [here](https://github.com/google/process.dart/blob/master/lib/src/interface/common.dart) ln 53. 

I can't imagine that it will actually run on my machine. it will only be executed on the custom device as far as I understand.

File: packages/flutter_tools/lib/src/artifacts.dart (hunk 2 above)
Comment (zanderso) [reply packages/flutter_tools/lib/src/artifacts.dart:957]: Sorry I'm missing some context, but I think you'd get a `gen_snapshot` under a `clang_arm64` directory in the build output if you're building on an `arm64` host. Is that the situation? That `gen_snapshot` should be targeting `arm64` but you can check by running `gen_snapshot --version`. It will show the target configuration in quotes at the end of the version string.

If you don't want the `gen_snapshot` under `.`, I'm curious what it is. Would you mind giving the output of `file gen_snapshot` and `gen_snapshot --version` for the `gen_snapshot` under `.`?

File: packages/flutter_tools/lib/src/artifacts.dart (hunk 3 above)
Comment (zanderso) [reply packages/flutter_tools/lib/src/artifacts.dart:971]: This could just use the file APIs to check that it exists.

File: packages/flutter_tools/lib/src/build_system/targets/custom_device.dart [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+import '../../build_info.dart';
+import '../../devfs.dart';
+import '../../project.dart';
+import '../build_system.dart';
+import '../depfile.dart';
+import '../exceptions.dart';
+import 'assets.dart';
+import 'common.dart';
+import 'icon_tree_shaker.dart';
+
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/build_system/targets/custom_device.dart:17]: nit: extra newline

File: packages/flutter_tools/lib/src/build_system/targets/custom_device.dart [hunk 6]
```diff
@@ -0,0 +1,171 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/build_system/targets/custom_device.dart:4]: nit: extra newline

File: packages/flutter_tools/lib/src/build_system/targets/custom_device.dart [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+  ];
+
+  @override
+  Future<void> build(Environment environment) async {
+    if (environment.defines[kBuildMode] == null) {
+      throw MissingDefineException(kBuildMode, 'bundle_custom_device_assets');
+    }
+    final BuildMode buildMode = getBuildModeForName(environment.defines[kBuildMode]!);
+    final Directory outputDirectory = environment.outputDir;
+      //.childDirectory('flutter_assets');
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/build_system/targets/custom_device.dart:60]: Remove commented code, please.

File: packages/flutter_tools/lib/src/artifacts.dart (hunk 4 above)
Comment (zanderso) [reply packages/flutter_tools/lib/src/artifacts.dart:168]: This should be whatever the output of the `flutter_engine_library` target is: https://github.com/flutter/engine/blob/main/shell/platform/embedder/BUILD.gn#L291.

That will be toolchain dependent. On Linux it'll be `libflutter_engine.so`.

File: packages/flutter_tools/lib/src/build_system/targets/custom_device.dart [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+    if (!outputDirectory.existsSync()) {
+      outputDirectory.createSync();
+    }
+
+    // Only copy the kernel blob in debug mode.
+    if (buildMode == BuildMode.debug) {
+      environment.buildDir.childFile('app.dill')
+        .copySync(outputDirectory.childFile('kernel_blob.bin').path);
+    }
+    environment.fileSystem.file(environment.artifacts.getArtifactPath(Artifact.libFlutterEngine)).copySync(outputDirectory.childFile('libflutter_engine.so').path);
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/build_system/targets/custom_device.dart:70]: Is it possible to avoid copying this string here and using a constant or a lookup using `Artifact.libFlutterEngine` instead?

File: packages/flutter_tools/lib/src/build_system/targets/custom_devices.dart [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+import '../depfile.dart';
+import '../exceptions.dart';
+import 'assets.dart';
+import 'common.dart';
+import 'desktop.dart';
+import 'icon_tree_shaker.dart';
+
+/// The only files/subdirectories we care out.
+const List<String> _kLinuxArtifacts = <String>[
+  'libflutter_linux_gtk.so',
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/build_system/targets/custom_devices.dart:20]: Now I'm super confused. This PR doesn't actually seem to be about generic support for profile builds for custom devices, but rather it's about something more specific. I think we need to take a step back from this PR, and at least I would like to have a better understanding of the problem being solved here. Sorry if that's already in an issue somewhere and I just missed it.

File: packages/flutter_tools/lib/src/artifacts.dart (hunk 4 above)
Comment (pieter-scholtz) [reply packages/flutter_tools/lib/src/artifacts.dart:168]: Thanks !

File: packages/flutter_tools/lib/src/artifacts.dart (hunk 1 above)
Comment (pieter-scholtz) [reply packages/flutter_tools/lib/src/artifacts.dart:20]: Noted.

File: packages/flutter_tools/lib/src/artifacts.dart (hunk 2 above)
Comment (pieter-scholtz) [reply packages/flutter_tools/lib/src/artifacts.dart:957]: ### For  `.`

**command:**
`/out/linux_debug_arm64$ file gen_snapshot`
 **output:**
`gen_snapshot: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, for GNU/Linux 3.7.0, BuildID[sha1]=c3db3f84ed98ba1c97b8af997bdfb2acb5231e7a, stripped
`

**command:**
`linux_debug_arm64$ ./gen_snapshot --version`
 **output:**
`bash: ./gen_snapshot: cannot execute binary file: Exec format error`

Do you need me to run this on arm64 to get the version , or is this enough info ?

### For `clang_x64`

**command:**
`out/linux_debug_arm64/clang_x64$ file gen_snapshot`
 **output:**
`gen_snapshot: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 3.2.0, BuildID[sha1]=04d12c6e9a3200644565d5c91cbd331cb8bc9a55, stripped
`

**command:**
`linux_debug_arm64/clang_x64$ ./gen_snapshot --version`
 **output:**
`Dart SDK version: 2.16.0-46.0.dev (dev) (Wed Nov 24 07:17:06 2021 -0800) on "linux_simarm64"
`

File: packages/flutter_tools/lib/src/build_system/targets/custom_devices.dart (hunk 9 above)
Comment (pieter-scholtz) [reply packages/flutter_tools/lib/src/build_system/targets/custom_devices.dart:20]: Sorry this is my mistake I copied the linux target , but did not modify them to the right extent. I will look into this.

File: packages/flutter_tools/lib/src/artifacts.dart (hunk 1 above)
Comment (pieter-scholtz) [reply packages/flutter_tools/lib/src/artifacts.dart:20]: Done.

File: packages/flutter_tools/lib/src/artifacts.dart (hunk 4 above)
Comment (pieter-scholtz) [reply packages/flutter_tools/lib/src/artifacts.dart:168]: I am a bit beat on how to solve this problem in a clean fashion, solutions I can think of :

1. Have a "hardcoded" lookup of some sort that returns different names based on the value of `platform (ln 160).

2. Have `engineLibraryFileName` as part of [custom_devices.json](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/static/custom-devices.schema.json) and use that to define the artifact file name.

3. A combination of the above options where we use 2 and have a "hardcoded" fallback or two.

File: packages/flutter_tools/lib/src/build_system/targets/custom_device.dart [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    if (!outputDirectory.existsSync()) {
+      outputDirectory.createSync();
+    }
+
+    // Only copy the kernel blob in debug mode.
+    if (buildMode == BuildMode.debug) {
+      environment.buildDir.childFile('app.dill')
+        .copySync(outputDirectory.childFile('kernel_blob.bin').path);
+    }
+    environment.fileSystem.file(environment.artifacts.getArtifactPath(Artifact.flutterEngineLibrary)).copySync(outputDirectory.childFile('libflutter_engine.so').path);
```
Comment (pieter-scholtz) [inline packages/flutter_tools/lib/src/build_system/targets/custom_device.dart:67]: Comment regarding copying went missing @zanderso : "Is it possible to avoid copying this string here and using a constant or a lookup using Artifact.libFlutterEngine instead?" 
```suggestion
    environment.fileSystem.file(environment.artifacts.getArtifactPath(Artifact.flutterEngineLibrary)).copySync(outputDirectory.childFile(basename(environment.artifacts.getArtifactPath(Artifact.flutterEngineLibrary))).path);
```

File: packages/flutter_tools/lib/src/artifacts.dart (hunk 3 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/artifacts.dart]: > This could just use the file APIs to check that it exists.

Yeah, something like `if (_fileSystem.file(libFlutterEnginePath).existsSync()) {`

File: packages/flutter_tools/lib/src/artifacts.dart (hunk 4 above)
Comment (pieter-scholtz) [reply packages/flutter_tools/lib/src/artifacts.dart:168]:
```suggestion
      r      switch (platform) {
        case TargetPlatform.windows_x64:
        case TargetPlatform.windows_uwp_x64:
          return 'flutter_engine.dll';
        case TargetPlatform.android:
        case TargetPlatform.android_arm:
        case TargetPlatform.android_arm64:
        case TargetPlatform.android_x64:
        case TargetPlatform.android_x86:
          throw StateError('Artifact $artifact not available for platform $platform.');
        case TargetPlatform.darwin:
          return 'libflutter_engine.dylib';
        case TargetPlatform.fuchsia_arm64:
        case TargetPlatform.fuchsia_x64:
        case TargetPlatform.ios:
          throw StateError('Artifact $artifact not available for platform $platform.');
        case TargetPlatform.linux_arm64:
        case TargetPlatform.linux_x64:
          return 'libflutter_engine.so';
        case TargetPlatform.tester:
        case TargetPlatform.web_javascript:
        case null:
          throw StateError('Artifact $artifact not available for platform $platform.');
      }
```

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 99.
View them at https://flutter-gold.skia.org/cl/github/96743

**Comment by christopherfujino** [comment]
This PR looks like it had a merge from a branch other than `master` resulting in a large diff and meaning it will not be mergeable. Please see [tree hygiene](https://github.com/flutter/flutter/wiki/Tree-hygiene) for more details about how to open pull requests to Flutter repositories.

**Comment by pieter-scholtz** [comment]
> This PR looks like it had a merge from a branch other than `master` resulting in a large diff and meaning it will not be mergeable. Please see [tree hygiene](https://github.com/flutter/flutter/wiki/Tree-hygiene) for more details about how to open pull requests to Flutter repositories.

This has been fixed AFAICT.

**Comment by christopherfujino** [comment]
Thank you for contributing! As you have not responded within 3 weeks, I will be closing this PR as stale. If you are still interested in landing this, feel free to rebase this PR upstream, address code review feedback, and re-open this PR.
